colcon_in_container/providers/multipass.py

```diff
@@ -0,0 +1,171 @@
+    """Multipass client interacting with the Multipass socket."""
+
+    def __init__(self, ros_distro):  # noqa: D107
+        super().__init__(ros_distro)
+
+        self.multipass_path = _get_multipass_path()
+        if not self.multipass_path:
+            raise exceptions.ProviderNotInstalledOnHostError(
+                'Multipass is not installed.'
+                'Please run `sudo snap install multipass`')
+
+        self._render_jinja_template()
+
+        if self._run(['info', self.instance_name]).returncode == 0:
+            self._clean_instance()
+
+        cpus = os.getenv('COLCON_IN_CONTAINER_MULTIPASS_CPUS', default='2')
+        mem = os.getenv('COLCON_IN_CONTAINER_MULTIPASS_MEMORY', default='2G')
+        disk = os.getenv('COLCON_IN_CONTAINER_MULTIPASS_DISK', default='256G')
+
+        logger.info('Downloading the image then '
+                    'creating the Multipass instance')
+        self._run(
+            ['launch', self.ubuntu_distro,
+             '--name', self.instance_name,
+             '--cpus', cpus,
+             '--memory', mem,
+             '--disk', disk,
+             '--cloud-init', self.rendered_cloud_init_path,
+             '--timeout', '1000'], check=True)
+        self.execute_command(['cloud-init', 'status', '--wait'])
+
+    def _render_jinja_template(self):
+        config_directory = os.path.join(
+            os.path.dirname(os.path.realpath(__file__)), 'config')
+        cloud_init_file = os.path.join(
+            config_directory, 'cloud-init.yaml')
+        with open(cloud_init_file, 'r') as f:
+            config = f.read()
+
+        template = jinja2.Environment().from_string(source=config)
+        cloud_init_content = template.render(
+            {'v1': {'machine': processor(),
+                    'distro_release': self.ubuntu_distro}}
+        )
+
+        self.rendered_cloud_init_path = '.colcon-in-container-cloud-init.yaml'
+        with open(self.rendered_cloud_init_path, 'w') as f:
+            written = f.write(cloud_init_content)
+            if not written:
+                raise exceptions.ProviderNotConfiguredError(
+                    'Failed to write cloud-init file '
+                    f'in {self.rendered_cloud_init_path}')
+
+    def _run(self, command: List[str], **kwargs):
+        """Execute a multipass command.
+
+        Returns the return code from the command.
+        """
+        command = [str(self.multipass_path), *command]
+
+        logger.debug(f'Executing on host: {" ".join(command)}')
+        return subprocess.run(command, **kwargs)
+
+    def _clean_instance(self):
+        self._run(['delete', '--purge', self.instance_name])
+
+    def execute_command(self, command: List[str]):
+        """Execute the given command inside the instance."""
+        completed_process = self._run(['exec', self.instance_name,
+                                       '--working-directory', '/ws',
+                                       '--', 'sudo', *command],
```

Comment:
'sudo' ?

Reply:
lxd containers spawn only with root user.
Multipass spawn with root and ubuntu user. By default, everything executed as the `ubuntu` user.
I decided to "ignore" the ubuntu user on multipass by running everything as root in the vm.

Reply:
Alright, we should try to use the regular user as much as possible. Let's leave it a that here but please do open a ticket for later.

Reply:
Ticket created: https://github.com/canonical/colcon-in-container/issues/18